### PR TITLE
Rangers

### PR DESCRIPTION
**Pull request by shengwenzheng**
**NOTE**

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Please enable "Allow edits by maintainers" while putting up the PR.
3. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
4. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
5. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
6. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
7. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Name (to be shown on DefiLlama):
Rangers

##### Twitter Link:
https://twitter.com/RangersProtocol
##### Website Link:
https://rangersprotocol.com/

##### Logo (High resolution, will be shown with rounded borders):
https://s2.coinmarketcap.com/static/img/coins/64x64/12221.png

##### Current TVL:
2.68M

##### Chain:
rpg

##### Category (full list at https://defillama.com/categories) *Please choose only one:
Dex

##### Oracle used (Chainlink/Band/API3/TWAP or any other that you are using):
Chainlink

##### forkedFrom (Does your project originate from another project):
https://github.com/DefiLlama/DefiLlama-Adapters

##### methodology (what is being counted as tvl, how is tvl being calculated):
  Count the sum of the five tokens (RPG, USDT, MIX, AMG, COIN) of the dapp platform (bridge, nodeManager) and Dehero contract on the rangers blockchain.

##### Github org/user (Optional, if your code is open source, we can track activity):yes

### CODE REVIEW

File: projects/ponytaswap/index.js [hunk 1]
```diff
@@ -1,11 +1,19 @@
 const { getUniTVL } = require('../helper/unknownTokens')
+const {staking} = require("../helper/staking");
 
+const pool = [
```
Comment (g1nt0ki) [inline projects/ponytaswap/index.js:4]: hi @shengwenzheng 
what are these pools? when I go to https://www.ponytaswap.finance/ I dont see anything in the pools/farms page. Bug in the UI?

File: projects/rangers/index.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+const tokens = [
+  '0x71d9cfd1b7adb1e8eb4c193ce6ffbe19b4aee0db',
+  '0x8e8816a1747fddc5f8b45d2e140a425d3788f659',
+  '0x36426b7bf5709e5c2160411c6e8b1832e3404fe1',
+  '0xdaa6a6919c9543d8787490f5e9ad532c4d7ce9e8',
+  '0x3978e3cAB1c503eFaD75cB929C7076B7f4f3b6F2'
+]
+const owners = [
+  '0x00852E9CDF414D6D9403468340f78ea93Eda82d9',
```
Comment (g1nt0ki) [inline projects/rangers/index.js:12]: similarly, I am unsure of all these addresses, like what they are, can you fill in the form in PR description so we get a better idea of this new listing.

Also, staking token value goes under 'staking'

**Comment by shengwenzheng** [comment]
Sorry, this is a mistake, I will delete the code later



At 2023-11-10 16:15:50, "g1nt0ki" ***@***.***> wrote:

@g1nt0ki commented on this pull request.

In projects/ponytaswap/index.js:

>  
+const pool = [


hi @shengwenzheng
what are these pools? when I go to https://www.ponytaswap.finance/ I dont see anything in the pools/farms page. Bug in the UI?

**Comment by shengwenzheng** [comment]
"Tokens" is the contract address of the five tokens on the rpg chain, "RPG, USDT, MIX, AMG, COIN". "tokens" is the contract address of the Dapp platform



At 2023-11-10 16:38:56, "g1nt0ki" ***@***.***> wrote:

@g1nt0ki commented on this pull request.

In projects/rangers/index.js:

> @@ -0,0 +1,36 @@
+const { sumTokens2 } = require('../helper/unwrapLPs');
+const {toUSDTBalances} = require('../helper/balances')
+
+const tokens = [
+  '0x71d9cfd1b7adb1e8eb4c193ce6ffbe19b4aee0db',
+  '0x8e8816a1747fddc5f8b45d2e140a425d3788f659',
+  '0x36426b7bf5709e5c2160411c6e8b1832e3404fe1',
+  '0xdaa6a6919c9543d8787490f5e9ad532c4d7ce9e8',
+  '0x3978e3cAB1c503eFaD75cB929C7076B7f4f3b6F2'
+]
+const owners = [
+  '0x00852E9CDF414D6D9403468340f78ea93Eda82d9',


similarly, I am unsure of all these addresses, like what they are, can you fill in the form in PR description so we get a better idea of this new listing.

Also, staking token value goes under 'staking'

File: projects/rangers/index.js (hunk 2 above)
Comment (shengwenzheng) [reply projects/rangers/index.js:12]: "Tokens" is the contract address of the five tokens on the rpg chain, "RPG, USDT, MIX, AMG, COIN". "tokens" is the contract address of the Dapp platform

File: projects/ponytaswap/index.js (hunk 1 above)
Comment (shengwenzheng) [reply projects/ponytaswap/index.js:4]: Sorry, this is a mistake, I will delete the code later

File: projects/rangers/index.js (hunk 2 above)
Comment (shengwenzheng) [reply projects/rangers/index.js:12]: "Tokens" is the contract address of the five tokens on the rpg chain, "RPG, USDT, MIX, AMG, COIN". "owners" is the contract address of the Dapp platform

**Comment by shengwenzheng** [comment]
"Tokens" is the contract address of the five tokens on the rpg chain, "RPG, USDT, MIX, AMG, COIN". "owners" is the contract address of the Dapp platform



At 2023-11-10 16:38:56, "g1nt0ki" ***@***.***> wrote:

@g1nt0ki commented on this pull request.

In projects/rangers/index.js:

> @@ -0,0 +1,36 @@
+const { sumTokens2 } = require('../helper/unwrapLPs');
+const {toUSDTBalances} = require('../helper/balances')
+
+const tokens = [
+  '0x71d9cfd1b7adb1e8eb4c193ce6ffbe19b4aee0db',
+  '0x8e8816a1747fddc5f8b45d2e140a425d3788f659',
+  '0x36426b7bf5709e5c2160411c6e8b1832e3404fe1',
+  '0xdaa6a6919c9543d8787490f5e9ad532c4d7ce9e8',
+  '0x3978e3cAB1c503eFaD75cB929C7076B7f4f3b6F2'
+]
+const owners = [
+  '0x00852E9CDF414D6D9403468340f78ea93Eda82d9',


similarly, I am unsure of all these addresses, like what they are, can you fill in the form in PR description so we get a better idea of this new listing.

Also, staking token value goes under 'staking'

**Comment by llamatester** [comment]
The adapter at projects/rangers exports TVL: 
        
 
 
    rpg                       2.68 M
    
    total                    2.68 M

**Comment by shengwenzheng** [comment]
Yes, how do I see it on the UI page?



At 2023-11-13 10:50:54, "llamatester" ***@***.***> wrote:

The adapter at projects/rangers exports TVL:

rpg                       2.68 M

total                    2.68 M

**Comment by g1nt0ki** [comment]
@shengwenzheng thanks for the PR but we are going to need more information if you want to list this project on defillama.

1. Please fill in the project details in the PR description

what the different contracts that you are checking token balance for and why should they be considered tvl?

**Comment by shengwenzheng** [comment]
hi @g1nt0ki
DeheroGame Intro:
DeHeroGame is a blockchain game published by MixMarvel with a global heroic fight as the background, and it was first released on Rangers Protocol. Rich hero development and combination of professions and skills, train different heroes from various parallel worlds. Reasonably allocate different heroes to challenge nearly a hundred dungeon levels of different difficulty, and start an adventure in a different world.

Rangers Connector Intro:
Rangers Connector is the best heterogeneous cross-chain solution leveraging the relay chain technology. It removes the cross-chain communication barriers and lowers the development difficulty for dApp developers in the Ethereum ecosystem. By solving the synchronization problem of heterogeneous data sources, Rangers Connector enables assets and data to transfer freely between chains, helps Web3 entrepreneurs increase NFT’s commercial value and information circulation, and sets a unified data-standard for complex blockchain applications.

Node Management Platform Intro:
Node management 2.0: Rangers Protocol's Node Management Platform 2.0 is an aggregated multi-functional tool to provide comprehensive support in node hosting. It offers node purchase and rental service and real-time data to assist users in becoming verification nodes and enjoying affordable and effortless mining. This latest version also integrated the delegated staking system for convenient RPG token delegation and reward claiming, bringing greater ease in navigating and a better user experience. Node Management Platform 2.0 is a one-stop solution for all users’ needs in node services and portfolio management.


The smart contracts we submit are all on the Rangers mainnet, from the above dapps. There are many types of smart contracts in the DeheroGame, including skill package contracts, NFT package contracts, Commission return contract and Hero Sale Contract.

The reason why they should be considered as the source of TVL in Rangers Protocol Eco system is that the dapps projects cannot run smoothly without the crypto asset (FT asset) on Rangers mainnet. The asset on the above contract represents the assets that the project needs to use to implement each feature in the product, not the data sources that Defilama officially labeled as not being available for TVL. In conclusion, we hope that the official Defilama team will help Rangers Protocol to update the above data, and if there is any other information that you need, please feel free to contact us, and we will be more than happy to provide it.



At 2023-11-14 00:27:08, "g1nt0ki" ***@***.***> wrote:

@shengwenzheng thanks for the PR but we are going to need more information if you want to list this project on defillama.

Please fill in the project details in the PR description

what the different contracts that you are checking token balance for and why should they be considered tvl?

**Comment by g1nt0ki** [comment]
hi @shengwenzheng 
I meant this for details: 
image


"The smart contracts we submit are all on the Rangers mainnet, from the above dapps. There are many types of smart contracts in the DeheroGame, including skill package contracts, NFT package contracts, Commission return contract and Hero Sale Contract."

We are bit strict on what we count as TVL, it has to be non protocol assets which users can withdraw from the contract. For example, user provides WRPG/USDT liquidity, we count the these tokens locked in protocol as tvl. But if user stakes Dehero NFT in your contract, it is not counted as tvl, nor do we count if user purchases NFTs with USDT and the USDT is locked in your contract since the user can never get back that USDT.

This is why I asked what each smart contract you listed relates to.

**Comment by llamatester** [comment]
The adapter at projects/rangers exports TVL: 
        
 
 
    rpg                       2.90 M
    
    total                    2.90 M

**Comment by shengwenzheng** [comment]
Hello, I submitted a PR the day before yesterday. Can I see the latest TVL data on Defillama now? (rangers)



At 2023-11-20 17:26:06, "g1nt0ki" ***@***.***> wrote:

hi @shengwenzheng
I meant this for details:


"The smart contracts we submit are all on the Rangers mainnet, from the above dapps. There are many types of smart contracts in the DeheroGame, including skill package contracts, NFT package contracts, Commission return contract and Hero Sale Contract."

We are bit strict on what we count as TVL, it has to be non protocol assets which users can withdraw from the contract. For example, user provides WRPG/USDT liquidity, we count the these tokens locked in protocol as tvl. But if user stakes Dehero NFT in your contract, it is not counted as tvl, nor do we count if user purchases NFTs with USDT and the USDT is locked in your contract since the user can never get back that USDT.

This is why I asked what each smart contract you listed relates to.

**Comment by g1nt0ki** [comment]
Hello, I submitted a PR the day before yesterday. Can I see the latest TVL data on Defillama now? (rangers)

No, we cant list it in its current state. You have not addressed any of my concerns with the code, given no clarifications

**Review by g1nt0ki — CHANGES_REQUESTED** [review]
.

**Comment by shengwenzheng** [comment]
hello,
The assets in the above contracts are the RPG liquidity that the users can provide, and user can withdraw or use it at any time.

According to the official documentation, the above assets are all part of "Staking" and "Pool2", which is a reasonable TVL, so please help RP to update the TVL data, thanks!



At 2023-11-23 16:56:34, "g1nt0ki" ***@***.***> wrote:

Hello, I submitted a PR the day before yesterday. Can I see the latest TVL data on Defillama now? (rangers)

No, we cant list it in its current state. You have not addressed any of my concerns with the code, given no clarifications

**Comment by g1nt0ki** [comment]
"The smart contracts we submit are all on the Rangers mainnet, from the above dapps. There are many types of smart contracts in the DeheroGame, including skill package contracts, NFT package contracts, Commission return contract and Hero Sale Contract."

We are bit strict on what we count as TVL, it has to be non protocol assets which users can withdraw from the contract. For example, user provides WRPG/USDT liquidity, we count the these tokens locked in protocol as tvl. But if user stakes Dehero NFT in your contract, it is not counted as tvl, nor do we count if user purchases NFTs with USDT and the USDT is locked in your contract since the user can never get back that USDT.

This is why I asked what each smart contract you listed relates to.

**Comment by llamatester** [comment]
The adapter at projects/rangers exports TVL: 
        
 
 
    rpg                       4.02 M
    
    total                    4.02 M

File: projects/rangers/index.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+//   '0xBf7C82A1B67B4Ed6095A12140F5cfc1Ae56783AF',
+//   '0xf060AbF8f9DFa4c266c4366feb6642a39c3B5a5e',
+//   '0x2aB030407955b9921D12dFAA2B928D6Ac30734Af',
+//   '0x20861b4Dc1A505d208321e6E8816824a0F4E4c94',
+//   '0xd7DFD2AB0BE47EE46Ca281859408A94aC466736E'
+// ]
+//the owners of rangers (dapp)
+const ownersRPG = {
+  bridge: "0x9c1CbFE5328DFB1733d59a7652D0A49228c7E12C",
+  nodeManager: "0x74448149F549CD819b7173b6D67DbBEAFd2909a7"
```
Comment (g1nt0ki) [inline projects/rangers/index.js:31]: hi  @shengwenzheng thanks for making the changes. We can list this as "RPG Bridge", this is a bridge between ethereum and RPG right? 
Can you skip "nodeManager" from tvl calculation, I am not what contract it is and as far as I can see, it holds no tokens. 

Regarding the bridge, can you use ethereum contract to count the bridged USDT on ethereum side instead on rpg chain side, we count the tokens on source chain in the bridges

**Comment by shengwenzheng** [comment]
***@***.***
yes,it is a bridge between ethereum and RPG。
"nodeManager" is one of our dapps. It has a contract, which contains USDT and RPG, which exist as balances. There is currently no balance, but it will be deposited in the future.
For other questions, I will discuss them with my development partners and answer you later.



At 2023-11-30 15:32:03, "g1nt0ki" ***@***.***> wrote:

@g1nt0ki commented on this pull request.

In projects/rangers/index.js:

> +//   '0xa0Ba72A2f8D3D619ffb70baAA7982c44708b145A',
+//   '0x0948C098bE91e61050c531686d344e0B8d804Fd3',
+//   '0xf29bDba5598a85e75D53edD2a880D30E477a5A9a',
+//   '0xA2Fad94B16675a6Aa9e89308cc0192a2C5F0749b',
+//   '0x99B4f686EF406C9c4FE5468655A2d30E266C5536',
+//   '0xb08439deE0Bc1E7D5945B590B178F6833dF0bea1',
+//   '0xBf7C82A1B67B4Ed6095A12140F5cfc1Ae56783AF',
+//   '0xf060AbF8f9DFa4c266c4366feb6642a39c3B5a5e',
+//   '0x2aB030407955b9921D12dFAA2B928D6Ac30734Af',
+//   '0x20861b4Dc1A505d208321e6E8816824a0F4E4c94',
+//   '0xd7DFD2AB0BE47EE46Ca281859408A94aC466736E'
+// ]
+//the owners of rangers (dapp)
+const ownersRPG = {
+  bridge: "0x9c1CbFE5328DFB1733d59a7652D0A49228c7E12C",
+  nodeManager: "0x74448149F549CD819b7173b6D67DbBEAFd2909a7"


hi @shengwenzheng thanks for making the changes. We can list this as "RPG Bridge", this is a bridge between ethereum and RPG right?
Can you skip "nodeManager" from tvl calculation, I am not what contract it is and as far as I can see, it holds no tokens.

Regarding the bridge, can you use ethereum contract to count the bridged USDT on ethereum side instead on rpg chain side, we count the tokens on source chain in the bridges

**Comment by llamatester** [comment]
The adapter at projects/rangers exports TVL: 
        
 
 
    rpg                       3.53 M
    
    total                    3.53 M

**Comment by shengwenzheng** [comment]
hello:
I deleted the USDT on the RPG chain side



At 2023-11-30 15:32:03, "g1nt0ki" ***@***.***> wrote:

@g1nt0ki commented on this pull request.

In projects/rangers/index.js:

> +//   '0xa0Ba72A2f8D3D619ffb70baAA7982c44708b145A',
+//   '0x0948C098bE91e61050c531686d344e0B8d804Fd3',
+//   '0xf29bDba5598a85e75D53edD2a880D30E477a5A9a',
+//   '0xA2Fad94B16675a6Aa9e89308cc0192a2C5F0749b',
+//   '0x99B4f686EF406C9c4FE5468655A2d30E266C5536',
+//   '0xb08439deE0Bc1E7D5945B590B178F6833dF0bea1',
+//   '0xBf7C82A1B67B4Ed6095A12140F5cfc1Ae56783AF',
+//   '0xf060AbF8f9DFa4c266c4366feb6642a39c3B5a5e',
+//   '0x2aB030407955b9921D12dFAA2B928D6Ac30734Af',
+//   '0x20861b4Dc1A505d208321e6E8816824a0F4E4c94',
+//   '0xd7DFD2AB0BE47EE46Ca281859408A94aC466736E'
+// ]
+//the owners of rangers (dapp)
+const ownersRPG = {
+  bridge: "0x9c1CbFE5328DFB1733d59a7652D0A49228c7E12C",
+  nodeManager: "0x74448149F549CD819b7173b6D67DbBEAFd2909a7"


hi @shengwenzheng thanks for making the changes. We can list this as "RPG Bridge", this is a bridge between ethereum and RPG right?
Can you skip "nodeManager" from tvl calculation, I am not what contract it is and as far as I can see, it holds no tokens.

Regarding the bridge, can you use ethereum contract to count the bridged USDT on ethereum side instead on rpg chain side, we count the tokens on source chain in the bridges

**Pull request merged**